### ISSUE TITLE
Update API endpoints

### ISSUE THREAD

**Issue by svilupp** [JuliaML/OpenAI.jl#50]
It would be excellent to add the newly available API endpoints and capabilities that have been announced today, eg,
[ ] Assistants (incl. Code Interpreter): https://platform.openai.com/docs/assistants/overview
[ ] Vision (with B64 encoding of images): https://platform.openai.com/docs/guides/vision
[ ] Audio (TTS/STT): https://platform.openai.com/docs/guides/text-to-speech
[ ] Moderations: https://platform.openai.com/docs/guides/moderation/quickstart
[ ] Files upload: https://platform.openai.com/docs/api-reference/files

Extending current API calls, eg,
[ ] JSON mode
[ ] Seed

Ideally, we would also add some simple examples to the docs. 
There are many cool capabilities (eg, function calling, logit-bias-classifiers, ...) that are easy to do with the existing APIs, but hard to access for beginners. 

I'm happy to take a stab at it over the weekend if no one is interested!

**Comment by rory-linehan** [comment]
I can plink away this weekend as well, maybe we can create a branch for every item listed to reduce stepping on toes

**Comment by svilupp** [comment]
Btw I’ll tackle gpt-4 vision first!

Also, I forgot to add the finetuning API - I’ll add that to the list.

**Comment by svilupp** [comment]
A few updates from my end, the functionality I've tried so far does not require any changes to the `create_chat` (unless we want to add nicer interfaces).

# Vision comprehension
(not image generation)
```
using OpenAI, Base64
base64_image = open("julia.png", "r") |> base64encode
hist = [
    Dict("role" => "user",
        "content" => [
            Dict("type" => "text", "text" => "Describe the provided image"),
            Dict("type" => "image_url",
                "image_url" => "data:image/png;base64,$(base64_image)")]),
]
resp = create_chat(api_key, "gpt-4-vision-preview", hist; max_tokens = 300)
resp.response[:choices][1][:message][:content]
# "This is the logo for the programming language Julia. The logo consists of the word \"julia\" written in lowercase black letters. Above the 'i' in \"julia,\" there are four colored circles in blue, green, red, and purple, arranged in a slight upward arc from left to right. The typeface is sans-serif and modern-looking, and the circles give the logo a playful and distinctive appearance. Julia is known as a high-level, high-performance programming language for technical computing."
```
Notice the array in the "content" key. 
We could add a kwarg `image_file, but it would require new dependency on Base64 and it would get complicated if there are multiple messages provided (ie, where would we inject the image_url).

I propose no action here (but some examples in the docs would be probably useful for newcomers!)

# JSON mode

... TBU... need to dash now but it works (will update later but wanted to avoid duplication)

# Seed model parameter

... TBU... need to dash now but it "kinds" works (will update later but wanted to avoid duplication)

# Logprobs model parameter

Adding `logprobs=5` should give you 5 tokens with the highest logprob (5 is the maximum allowed, I guess to prevent distillation?).
However, it is not available yet (11th Nov).

**Comment by rory-linehan** [comment]
Catching up after having no time the past week..

I'm going to be fixing [this issue](https://github.com/JuliaML/OpenAI.jl/issues/37) first this weekend, but after that I can start taking a look at other improvements and these new features.

As far as JSON mode, what do you think of [this](https://github.com/JuliaML/OpenAI.jl/pull/52)? It seems to be at least tangentially related.

BTW, I'm all ears (and approvals :laughing: ) on improvements, nicer interfaces, etc. from power users like yourself. I favor a laissez-faire approach in these initial stages, so feel free to open PRs and we can discuss.

**Comment by svilupp** [comment]
I’ve commented on the above mentioned thread. I personally know very little about the streaming use case - everything I do doesn’t need it. I’m familiar with it only for the ChatGPT-like interfaces.
 
As far as the nicer interfaces go, I am not sure I have any ideas. I think what you’ve done is actually great and allows everyone who wants to build on top of OpenAI API to get started.

Based on my experience and talking to a few people, I think the issue now is awareness and how-tos. IMO, we need to focus on surfacing practical applications and lowering the barrier to entry for first time users, eg, examples, blogs, and potentially also some downstream libraries that can be very opinionated and focus on specific tasks.

That’s why I have bundled up bunch of scripts I had and I wrapped them in a library: [PromptingTools.jl](https://github.com/svilupp/PromptingTools.jl). The hope is to abstract and re-use prompts and be backend-agnostic (eg, switch between Open AI API or Mistral depending on your needs/tasks). I’m targeting tge daily “mini-tasks” that we all have and don’t enjoy.
I have a few more PRs locally but then I hope to change over to churn out bunch of how-to guides. The coverage of GenAI applications in Julia is sooooo poor right now.

You’ll see that some of the Issues I opened today reflect the above beliefs.

I’m keen to add the above APIs, but I don’t have any mini-tasks that need them right now, so I might not get to it for 1-2 weeks.

**Comment by rory-linehan** [comment]
Agreed on documentation and lowering the barrier to entry, I updated the readme yesterday with an example of overriding the base URL.

I'm starting on the Files item above, then Finetuning because the latter is dependent on the former.

**Comment by cpfiffer** [comment]
#57 tackles the assistant stuff. I'll leave the file stuff out of that PR until the Files API is implemented.
